101 hub and spoke sandbox

vmehmeri opened the pull request:
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [X] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-quickstart-templates?pullRequest=5040) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-quickstart-templates?pullRequest=5040) before we can accept your contribution.<br/><br/>:x: vmehmeri [sign now](https://cla.opensource.microsoft.com/Azure/azure-quickstart-templates?pullRequest=5040)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-quickstart-templates?pullRequest=5040) it.</sub>

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+      "winVmSize":{
+         "type":"string",
+         "defaultValue":"Standard_D3_v2",
+         "allowedValues":[
```
remove allowedValues since it will create a maint problem later

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+      },
+      "winVmUser":{
+         "type":"string",
+         "defaultValue":"WinAdmin",
```
remove defaultValue - increases surface area

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+      },
+      "winVmDns":{
+         "type":"string",
+         "defaultValue":"hubwinjumphostvm",
```
remove defaultValue here since it's not globally unique

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+         "type":"Microsoft.Network/virtualNetworks",
+         "name":"[variables('hubVnetName')]",
+         "apiVersion":"2018-02-01",
+         "location":"[resourceGroup().location]",
```
all of these must be parameters('location')

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+         "type":"Microsoft.Network/virtualNetworks",
+         "name":"[variables('devSpokeVnetName')]",
+         "apiVersion":"2018-02-01",
+         "location":"[resourceGroup().location]",
```
must be parameters('location')

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+         "name":"[variables('hubVnetName')]",
+         "apiVersion":"2018-02-01",
+         "location":"[resourceGroup().location]",
+         "scale":null,
```
remove this

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+         "type":"Microsoft.Network/virtualNetworks/subnets",
+         "name":"[concat(variables('hubVnetName'), '/', variables('mgmtSubnetName'))]",
+         "apiVersion":"2018-02-01",
+         "scale":null,
```
remove this

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+         "type":"Microsoft.Network/virtualNetworks/subnets",
+         "name":"[concat(variables('hubVnetName'), '/', variables('sharedSubnetName'))]",
+         "apiVersion":"2018-02-01",
+         "scale":null,
```
remove this

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+         "type":"Microsoft.Network/virtualNetworks/subnets",
+         "name":"[concat(variables('hubVnetName'), '/', variables('dmzSubnetName'))]",
+         "apiVersion":"2018-02-01",
+         "scale":null,
```
remove this

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+         "type":"Microsoft.Network/virtualNetworks/subnets",
+         "name":"[concat(variables('hubVnetName'), '/', variables('gatewaySubnetName'))]",
+         "apiVersion":"2018-02-01",
+         "scale":null,
```
remove this

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+         "name":"[variables('devSpokeVnetName')]",
+         "apiVersion":"2018-02-01",
+         "location":"[resourceGroup().location]",
+         "scale":null,
```
remove this

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+                  "[variables('devSpokeVnetPrefix')]"
+               ]
+            },
+            "dhcpOptions":{
```
remove empty properties/objects

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+                  "[variables('hubVnetPrefix')]"
+               ]
+            },
+            "dhcpOptions":{
```
remove all the empty objects/properties here and below

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+         "type":"Microsoft.Network/virtualNetworks/subnets",
+         "name":"[concat(variables('devSpokeVnetName'), '/', variables('spokeWorkloadSubnetName'))]",
+         "apiVersion":"2018-02-01",
+         "scale":null,
```
remove all of these "null" properties (won't comment on the rest but need to remove them)

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+            "diagnosticsProfile":{
+               "bootDiagnostics":{
+                  "enabled":true,
+                  "storageUri":"[concat('https', '://', variables('winVmStorageName'), '.blob.core.windows.net', '/')]"
```
See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#referencing-resource-properties) for how to handle storage account uris

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.json:
```diff
@@ -0,0 +1,916 @@
+            "idleTimeoutInMinutes":4,
+            "dnsSettings":{
+               "domainNameLabel":"[toLower(parameters('winVmDns'))]",
+               "fqdn":"[concat(toLower(variables('winJmphostName')), '.[resourceGroup().location].cloudapp.azure.com')]"
```
remove this

bmoore-msft commented on 101-hub-and-spoke-sandbox/azuredeploy.parameters.json:
```diff
@@ -0,0 +1,30 @@
+{
```
remove all non-GEN values from here - use the defaultValue in the template

vmehmeri replied:
Done and acted on all other changes as well

vmehmeri commented:
Acted on requested changes:
- Removed empty properties and null properties
- Removed default and allowed values where it could cause problems
- Used parameters('location') instead of resourceGroup().location on resources
- Fixed storage account URI according to guideline
- Removed all non-GEN variables from parameters file.

Also added dependency between subnet resources of the hub VNet, to avoid a conflict issue that would occur on occasion during deployment (subnets deployed at the same time in the same VNet - AnotherOperationInProgress error). 

Rename WinVmDns variable to WinVmDnsPrefix
